### ISSUE TITLE
Weird non-highlighting output

### ISSUE THREAD

**Issue by petdance** [beyondgrep/ack3#101]
````
$ cat foo
            TWise::TCheck::Constants::TITLEONLY_T
            TWise::TCheck::Constants::POSSIBLE_TA
            TWise::TCheck::Constants::TITLEONLY_T
            TWise::TCheck::Constants::POSSIBLE_TP
````

Then:

````
$ ack3 '[A-Z][a-z]+::[A-Z_]+[^:\w]'  foo 
            TWise::TCheck::Constants::TITLEONLY_T
            TWise::TCheck::Constants::POSSIBLE_TA
            TWise::TCheck::Constants::TITLEONLY_T
            TWise::TCheck::Constants::POSSIBLE_TP
````

The lines that come out are not highlighted at all.  Why not?  `ack2` does the same thing.

**Comment by n1vux** [comment]
`--output='$&'  `  further reduces the output , gets nothing !

**Comment by n1vux** [comment]
My suspicion is the `[^:\w]` should not be matching  at all, so highlight re-search doesn't find a match for `$&`, but the line was erroneously accepted as matching.   
I'm guessing you were expecting the same highlight as with  '[A-Z][a-z]+::[A-Z_]+$'  e.g. Constants to end of line.  I don't thing the original pattern should match these lines, some bug has it matching them but not highlighting, and i suspect that is due to chomping or if we allow [^] to match newline, do we have //s on one match but not the other ?

**Comment by n1vux** [comment]
yes, `'[A-Z][a-z]+::[A-Z_]+[\n]' ` matches the line but no highlight, while `'[A-Z][a-z]+::[A-Z_]+[\n]'` highlights

**Comment by n1vux** [comment]
do we match line before chomping but chomp before matching highlight?   Even a zero with lookahead for a newline fails to highlight `
`'[A-Z][a-z]+::[A-Z_]+(?=\s)'`

**Comment by n1vux** [comment]
Shortest whittle 
`ack3 '.*\n$'  foo`

This may be a doc issue.

- if caring about highlights, don't match the newline; 
- use a boundary `\b` `$`  `\Z`  (but not `\z`)

OTOH, we might want to chomp off trailing `[\r\n]+$` before doing the main match to prevent `\W` or `[^:\w]` matching a `\n` that we can't find for highlighting.

**Comment by petdance** [comment]
Probably we need to start chomping on reading the line.  The `\n` causes us no end of problems.

**Comment by n1vux** [comment]
(cross ref  #102 on that topic)
